app/Http/Controllers/Api/VolumeController.php

```diff
@@ -184,213 +158,304 @@ public function update(UpdateVolume $request)
+     * @apiSuccessExample {json} Success response:
+     * {
+     * "name": "Kulas Group",
+     * "media_type_id": 3,
+     * "creator_id": 5,
+     * "url": "test://files",
+     * "handle": null,
+     * "created_at": "2022-11-25T13:10:12.000000Z",
+     * "updated_at": "2022-11-25T13:10:12.000000Z",
+     * "id": 4
+     * }
+     **/
     public function clone($volumeId, $destProjectId, Request $request)
     {
         return DB::transaction(function () use ($volumeId, $destProjectId, $request) {
             $project = Project::findOrFail($destProjectId);
             $this->authorize('update', $project);
             $volume = Volume::findOrFail($volumeId);
+            $this->authorize('update', $volume);
             $copy = $volume->replicate();
             $copy->name = $request->input('name', $volume->name);
             $copy->created_at = Carbon::now();
             $copy->save();
 
-            if ($volume->mediaType->name == "image") {
+            if ($volume->isImageVolume()) {
                 $this->copyImages($volume, $copy, $request->input('imageIds', []));
-                $this->copyColorSortSequence($volume, $copy);
+                $this->copyImageAnnotation($volume, $copy, $request->input('imageIds', []));
+                $this->copyImageLabels($volume, $copy, $request->input('imageIds', []));
             } else {
                 $this->copyVideos($volume, $copy, $request->input('videoIds', []));
+                $this->copyVideoAnnotation($volume, $copy, $request->input('videoIds', []));
+                $this->copyVideoLabels($volume, $copy, $request->input('videoIds', []));
             }
-            $this->copyAnnotationSessions($volume, $copy);
 
             //save ifdo-file if exist
             if ($volume->hasIfdo()) {
                 $this->copyIfdoFile($volumeId, $copy->id);
             }
 
-            $copy->push();
-
             $project->addVolumeId($copy->id);
-            $project->push();
 
-            return redirect()->back()->with(['copy' => $copy]);
+            if ($this->isAutomatedRequest()) {
+                return $volume;
+            }
+
+            return $this->fuzzyRedirect()
+                ->with(["copy" => $copy])
+                ->with('messageType', 'success');
         });
 
     }
 
-    //TODO: add javadocs
-    private function copyImages($volume, $copy, $imageIds)
+    /**
+     * Copies (selected) images from given volume to volume copy.
+     *
+     * @param Volume $volume
+     * @param Volume $copy
+     * @param int[] $selectedImageIds
+     **/
+    private function copyImages($volume, $copy, $selectedImageIds)
     {
         // copy image references
-        $images = count($imageIds) == 0 ? $volume->images()->get() : $volume->images()->whereIn('id', $imageIds)->get();
-        $images = $images->map(function ($image) use ($copy) {
+        $images = $volume->images()
+            ->when(!empty($images), function ($query) use ($selectedImageIds) {
```

Comment:
My suggestion would be to check the type of $selectedImageIds. If it is an array then use !empty() otherwise isEmpty(). But maybe you know a more elegant solution?

Reply:
I just ran a test and it seems `$request->input('imageIds', [])` always returns a plain array. So no type check needed and you can just use `empty()`.